### PR TITLE
Grouped box- and violin plots

### PR DESCRIPTION
**Pull request by kescobo**
Goal to address #145 . Testing with:

```
using Random
using DataFrames
using StatPlots

df = DataFrame(x=repeat([1,2,3],4), y=rand(12), g=vcat(fill("a", 6), fill("b", 6))
    )

@df df groupedbar(:x, :y, group=:g)
```
screen shot 2018-10-11 at 14 23 45

```
@df df boxplot(:x, :y, group=:g)
```
screen shot 2018-10-11 at 14 24 38

```
@df df violin(:x, :y, group=:g)
```
screen shot 2018-10-11 at 14 25 04

EDIT: this also addresses #56 and #89

### CODE REVIEW

File: src/boxplot.jl [hunk 1]
```diff
@@ -21,6 +21,25 @@ notch_width(q2, q4, N) = 1.58 * (q4-q2)/sqrt(N)
     bw == nothing && (bw = 0.8)
     @assert whisker_width == :match || whisker_width >= 0 "whisker_width must be :match or a positive number"
     ww = whisker_width == :match ? bw : whisker_width
+
+    nr, nc = size(y) # size(y) == (6,)
```
Comment (kescobo) [inline src/boxplot.jl]: I copied this code from box.jl, but it's giving me an error at the moment because the return is a 1-element tuple.

I tried adding in the `grouped_xy()` function from `bar.jl` as well, but that doesn't seem to work. I'm guessing the `x` and `y` pulled in by `g.args...` in:

```
@recipe function f(g::GroupedBar)
    x, y = grouped_xy(g.args...)
     ...
```

Is different than from `@recipe function f(::Type{Val{:boxplot}}, x, y, z;...)`, but I don't understand enough to know how to fix.

**Comment by mkborregaard** [comment]
Can you post the image?

**Comment by kescobo** [comment]
@mkborregaard Which image? The ones from the current default?

**Comment by mkborregaard** [comment]
Yes :-)

**Comment by kescobo** [comment]
Done! I think I just need to know how to get the number of groups from the `plot_attributes`?

**Comment by kescobo** [comment]
I'm a bit lost here - I just hard-coded in 2 groups for now, but I can't figure out how this would work. It looks like the series are being generated separately? So even if I can find the number of groups, I don't know how I would write code to use the group number to calculate an offset.

Eg, if I do:

```
    ...
    loops = 0
    for (i,glabel) in enumerate(glabels)
        loops += 1
        @info loops
    ...
```
The output is
```
[ Info: 1
[ Info: 2
[ Info: 3
[ Info: 1
[ Info: 2
[ Info: 3
```

**Comment by mkborregaard** [comment]
It's been so long I touched the grouping machinery I don't remember it anymore. You should be able to get it directly with `length(unique(g))`, though, right?

**Comment by kescobo** [comment]
Woo! 

```
@df df boxplot(:x, :y, group=:g)
```
screen shot 2018-10-15 at 11 02 05


and

```
@df df boxplot(:g, :y, group=:x)
```
screen shot 2018-10-15 at 11 02 17

**Comment by kescobo** [comment]
Still not done - working on violin now

**Comment by kescobo** [comment]
screen shot 2018-10-15 at 11 07 26


screen shot 2018-10-15 at 11 07 44

**Comment by kescobo** [comment]
Is this what we expect:

```
@df df boxplot(:x, :y, group=:g, bar_width=0.4)
@df df violin(:x, :y, group=:g, bar_width=0.4)
```

screen shot 2018-10-15 at 11 10 05

screen shot 2018-10-15 at 11 10 11

Is there a way to add space between the bars that I'm missing? If not, I think this is ready (other than tests, which I have zero idea how to write).

**Comment by piever** [comment]
I think tests on StatPlots plots were moved to Plots and then disabled, so for the time being there's no need to worry so much about them. Here the space from one `x` value to the next seems a bit excessive though, I personally prefer the one from the image here https://github.com/JuliaPlots/StatPlots.jl/pull/192#issuecomment-429890341

To add space for the different groups (say from the blue and orange) you can cheat a bit and reduce the `bw` of an additional say 20% (it's how we do for bar plots as far as I remember: `bw = 0.8/bw`). Though 20% is the distance across separate "chunks" so maybe this one could be a bit less extreme, maybe 10% (or `sqrt(20)`% if somehow a geometric progression is more pleasant to look at). At the same time here it looks like the space across chunks is much bigger than 20% of the chunk, not sure how that changed.

**Comment by kescobo** [comment]
>I personally prefer the one from the image here #192 (comment)

Yeah - that's the default. The latter is with `bar_width` set to a smaller amount.

File: src/boxplot.jl [hunk 2]
```diff
@@ -21,10 +22,20 @@ notch_width(q2, q4, N) = 1.58 * (q4-q2)/sqrt(N)
     bw == nothing && (bw = 0.8)
     @assert whisker_width == :match || whisker_width >= 0 "whisker_width must be :match or a positive number"
     ww = whisker_width == :match ? bw : whisker_width
+
+    ngroups = length(unique(plotattributes[:group]))
+    thisgroup = plotattributes[:series_plotindex]
+
+    outercenter = 0.5 * bw - bw / 2ngroups
+    offsets = collect(-outercenter:(bw / ngroups):outercenter)
+    bw = bw / ngroups
```
Comment (piever) [inline src/boxplot.jl]: Maybe consider:

```julia
bw = bw / ngroups * 0.9
ww = ww / ngroups * 0.9
```

to allow for some space within boxes of the same group.

Comment (kescobo) [reply src/boxplot.jl]: I think having the bars touching (as in `groupedbar`) makes sense - I was just wondering if there would be an appropriate keyword to grab to adjust that.

FWIW, this is what your suggestion looks like:

screen shot 2018-10-15 at 15 05 10
screen shot 2018-10-15 at 15 05 00

**Comment by kescobo** [comment]
Actually, hold off on this - the `scatter` for the outliers isn't adjusted yet, and I've got some stray logging messages

**Comment by kescobo** [comment]
Actually, it's ok I think, but there's something weird about how colors are selected for the scatter. But that's a separate issue as far as I can tell. New example code:

```
using Random
using DataFrames
using StatPlots

function rand_outlier(magnitude::Int=3)
    v = rand()
    return rand([-magnitude, magnitude]) * v
end

function grouped_df(n::Int, nx::Int, ng::Int, ol::Int)
    return DataFrame(x=repeat(collect(1:nx), ng*n),
               y=[rand() < ol ? rand_outlier() : rand() for _ in 1:n*ng*nx],
               g=vcat([fill(randstring(), nx*n) for _ in 1:ng]...))
end
```

```
@df grouped_df(100, 3, 2, 0.1) boxplot(:x, :y, group=:g, legend=false)
```

image


```
@df grouped_df(100, 3, 1, 0.1) boxplot(:x, :y, group=:g, legend=false)
```

image


```
@df grouped_df(100, 5, 3, 0.1) boxplot(:x, :y, group=:g, legend=false, color = [:red :green :blue])
```

image

**Comment by kescobo** [comment]
@piever @mkborregaard I think this is ready? Opening a separate issue for the outlier thing.

**Comment by kescobo** [comment]
Uggggh... hold on. Now it doesn't work _without_ groups. 

```
julia> @df grouped_df(100, 3, 1, 0.1) boxplot(:x, :y, legend=false)
ERROR: MethodError: no method matching iterate(::Nothing)
Closest candidates are:
  iterate(::Core.SimpleVector) at essentials.jl:578
  iterate(::Core.SimpleVector, ::Any) at essentials.jl:578
  iterate(::ExponentialBackOff) at error.jl:171
  ...
```

**Comment by kescobo** [comment]
OK, ready for real this time

**Comment by mkborregaard** [comment]
Hi this is awesome - but how do you get a space between the boxes in a group? To me it looks weird that they touch.

File: src/boxplot.jl (hunk 2 above)
Comment (piever) [reply src/boxplot.jl]: The last thing is whether we should get an attribute to avoid "touching" boxplots - and what the default should be. Maybe they can default with no distance and `boxplot(...., space = 0.1)` could create some space between them (like in the example above, with `bw = bw / ngroups * (1-space)`.

@mkborregaard: what do you think is a sensible name for the attribute and a sensible default?

**Comment by piever** [comment]
I commented above on the "touching". I agree this is awesome btw, it fixes 1/3 of the issues of StatPlots :)

**Comment by kescobo** [comment]
>how do you get a space between the boxes in a group? To me it looks weird that they touch.

We could hardcode something ([as I did here](https://github.com/JuliaPlots/StatPlots.jl/pull/192#discussion_r225282294)), but I'm not sure what keyword would make sense to specify this. At the moment, the bars in `groupedbar` are touching, so that's how I coded it. 

FWIW - that's how it works in `ggplot2` as well: https://www.r-graph-gallery.com/265-grouped-boxplot-with-ggplot2/

**Comment by piever** [comment]
Given that `groupedbar` is touching and in `ggplot2` it's touching as well, I think this could be merged as is (touching) and we can later think what's a good keyword to specify the distance. It should probably be done in a separate PR as the same keyword should, IMO, also work for `groupedbar`

**Comment by kescobo** [comment]
@piever Sounds good to me. I don't have anything to add at this point

**Comment by kescobo** [comment]
I added an issue to keep track of that idea, but I think this PR is gtg :shipit:

File: src/boxplot.jl (hunk 2 above)
Comment (mkborregaard) [reply src/boxplot.jl]: `spacing`? I think sth like 0.1 should be the default.

**Comment by mkborregaard** [comment]
Great, thanks a lot @kescobo !

**Comment by Ibrahim-a-Ibrahim** [comment]
> Woo!
> 
> ```
> @df df boxplot(:x, :y, group=:g)
> ```
> 
> screen shot 2018-10-15 at 11 02 05
> 
> and
> 
> ```
> @df df boxplot(:g, :y, group=:x)
> ```
> 
> screen shot 2018-10-15 at 11 02 17

I tried the following code and it sounds that it does not result any grouping.
```
using Plots
using StatsPlots

x = rand(2,24,3)
x1 = x[:,:,1]
```
groupedboxplot(x, bar_position = :dodge, bar_width=0.7)
groupedboxplot(x1, bar_position = :dodge, bar_width=0.7)
Any suggestion to end up with a grouped boxplot?

**Comment by mkborregaard** [comment]
StatsPlots expect data to be "long" format, e.g. https://r4ds.had.co.nz/tidy-data.html

**Comment by Ibrahim-a-Ibrahim** [comment]
> StatsPlots expect data to be "long" format, e.g. https://r4ds.had.co.nz/tidy-data.html

Could you please help to provide a piece of code to represent the issue that I raised it up using the "long" format data frame?

**Comment by mkborregaard** [comment]
What do you expect the groups to be? You data has no categorical variables to group on

**Comment by Ibrahim-a-Ibrahim** [comment]
> What do you expect the groups to be? You data has no categorical variables to group on

I am looking to plot the data from the 3D array like the attached photo. Is it helpful?
3phase

**Comment by mkborregaard** [comment]
Not really - you are taking the `group` information from the column it is in, but the `x` variable is also not a grouped variable. You are looking for something like

```julia
using StatsPlots

x = rand(1:2, 100)
gr = rand(1:3, 100)
y = rand(100)

groupedboxplot(x, y, group = gr)
```

Skærmbillede 2020-08-19 kl  12 44 11

Although in this case, `groupedboxplot` probably should not be using numeric ticks.

**Comment by kescobo** [comment]
Also - doesn't really make sense to have this discussion in a merged PR. Slack, zulip, or discourse are probably more appropriate.

**Comment by kescobo** [comment]
Ah, I see you did that [here](https://discourse.julialang.org/t/grouped-boxplot-https-github-com-juliaplots-statsplots-jl-pull-192/44655/7?u=kevbonham), extra doesn't make sense to renew this old PR then (which was superceded by #211 in any case)
